py4cast/io/writing_outputs.py

```diff
@@ -0,0 +1,377 @@
+    leadtimes = validtimes - init_term
+    predicted_time_steps = len(leadtimes)
+
+    model_ds = {
+        c: xr.open_dataset(
+            Path(saving_settings.directory) / saving_settings.template_grib,
+            backend_kwargs={
+                "indexpath": "",
+                "read_keys": [
+                    "level",
+                    "shortname",
+                    "centre",
+                    "typeOfGeneratingProcess",
+                    "generatingProcessIdentifier",
+                    "typeOfLevel",
+                    "discipline",
+                    "parameterCategory",
+                    "parameterNumber",
+                    "unit",
+                ],
+                "filter_by_keys": grib_groups[c],
+            },
+        )
+        for c in grib_groups.keys()
+    }
+
+    for t_idx in range(predicted_time_steps)[:1]:
+        for group in model_ds.keys():
+            target_ds = deepcopy(model_ds[group])
+
+            # if the shape of the dataset grid doesn't match grib template, fill the rest of the data with NaNs
+            nanmask, latlon = make_nan_mask(ds, target_ds)
+            (
+                latmin,
+                latmax,
+                longmin,
+                longmax,
+            ) = latlon
+
+            target_ds["time"] = sample.date
+            ns_step = np.timedelta64(
+                int(leadtimes[t_idx] * 3600 * 1000000000),
+                "ns",
+            )
+            ns_valid = np.timedelta64(
+                int(validtimes[t_idx] * 3600 * 1000000000),
+                "ns",
+            )
+            target_ds["step"] = ns_step
+            target_ds["valid_time"] = np.datetime64(sample.date) + ns_valid
+
+            for feature_name in pred.feature_names_to_idx.keys():
+
+                level, name, tol = grib_features.loc[
+                    feature_name, ["level", "name", "typeOfLevel"]
+                ]
+
+                if (
+                    (f"{name}_{tol}" == group)
+                    or (f"{level}_{tol}" == group)
+                    or (tol == group)
+                ):
+                    data = (
+                        (
+                            pred.tensor[
+                                0, t_idx, :, :, pred.feature_names_to_idx[feature_name]
+                            ]
+                            .cpu()
+                            .numpy()
+                            .astype(np.float32)
+                        )
+                        # TODO : correctly reshape spatial dims in the 1D-catch-all case
+                        if pred.num_spatial_dims == 2
+                        else (
+                            pred.tensor[
+                                0, t_idx, :, pred.feature_names_to_idx[feature_name]
+                            ]
+                            .cpu()
+                            .numpy()
+                            .astype(np.float32)
+                        )
+                    )
```

Comment:
As it is your code makes assumptions about dimension ordering. You could use the "new" features of the NamedTensors to get rid off that.

```python
# collapse batch dim
pred.squeeze_("batch")

# extract a given feature at a given timestep
pred.select_dim("timestep", t_idx, bare_tensor=False)["feature_name"]
```

Reply:
done, waiting for PR 63

Reply:
still getting an error b'cause of the unsqueeze_() in getitem